osquery/remote/http/http_client.cpp

```diff
@@ -0,0 +1,342 @@
+          rc);
+
+  if (rc) {
+    std::string error("Failed to connect to ");
+    if (client_options_.proxy_hostname_) {
+      error += "proxy host ";
+    }
+    error += connect_host + ":" + port;
+    throw std::system_error(
+        std::error_code(rc.value(), std::generic_category()), error);
+  }
+
+  if (client_options_.proxy_hostname_) {
+    std::string remote_host = *client_options_.remote_hostname_;
+    std::string remote_port = *client_options_.remote_port_;
+
+    beast_http_request req;
+    req.method(beast_http::verb::connect);
+    req.target(remote_host + ":" + remote_port);
+    req.version = 11;
+    req.prepare_payload();
+    beast_http::write(sock_, req);
+
+    boost::beast::flat_buffer b;
+    beast_http_response_parser rp;
+    rp.skip(true);
+    beast_http::read_header(sock_, b, rp);
+    if (rp.get().result() != beast_http::status::ok) {
+      throw std::runtime_error(rp.get().reason().data());
+    }
+  }
+}
+
+void Client::encryptConnection() {
+  boost_asio::ssl::context ctx{boost_asio::ssl::context::sslv23};
+
+  if (client_options_.always_verify_peer_) {
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+  } else {
+    ctx.set_verify_mode(boost_asio::ssl::verify_none);
+  }
+
+  if (client_options_.server_certificate_) {
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+    ctx.load_verify_file(*client_options_.server_certificate_);
+  }
+
+  if (client_options_.verify_path_) {
+    ctx.set_verify_mode(boost_asio::ssl::verify_peer);
+    ctx.add_verify_path(*client_options_.verify_path_);
+  }
+
+  if (client_options_.ciphers_) {
+    ::SSL_CTX_set_cipher_list(ctx.native_handle(),
+                              client_options_.ciphers_->c_str());
+  }
+
+  if (client_options_.ssl_options_) {
+    ctx.set_options(client_options_.ssl_options_);
+  }
+
+  if (client_options_.client_certificate_file_) {
+    ctx.use_certificate_file(*client_options_.client_certificate_file_,
+                             boost_asio::ssl::context::pem);
+  }
+
+  if (client_options_.client_private_key_file_) {
+    ctx.use_private_key_file(*client_options_.client_private_key_file_,
+                             boost_asio::ssl::context::pem);
+  }
+
+  ssl_sock_ = std::make_shared<ssl_stream>(sock_, ctx);
+  if (client_options_.sni_hostname_) {
+    ::SSL_set_tlsext_host_name(ssl_sock_->native_handle(),
+                               client_options_.sni_hostname_->c_str());
+  }
+
+  boost_system::error_code rc;
+  ssl_sock_->handshake(boost_asio::ssl::stream_base::client, rc);
+
+  if (rc) {
+    throw std::system_error(rc.value(), adapted_category(&rc.category()));
+  }
+}
+
+template <typename STREAM_TYPE>
+void Client::sendRequest(STREAM_TYPE& stream,
+                         Request& req,
+                         beast_http_response_parser& resp) {
+  req.target((req.remotePath()) ? *req.remotePath() : "/");
+  req.version = 11;
+  req << Request::Header("Host", *client_options_.remote_hostname_);
```

Comment:
I'm struggling to see the value of this syntax, instead of:
```
req.set(beast_http::field::host, *client_options_.remote_hostname_);
```
The less code, the better, `Request::Header` seems unnecessary.

Reply:
Done. Within the boundaries of http_client , I can use the direct calls of beast.